contrib/hamilton/contrib/user/skrawcz/fine_tuning/README.md

```diff
@@ -0,0 +1,130 @@
+
+- Data. The data set should be list of JSON objects, where each entry in the list is an object that has the "question" and
+and also the "answer" to that question. You will provide the name of the keys for these fields as input to run the code.
+e.g. you should be able to do `json.load(f)` and it would return a list of dictionaries, e.g. something like this:
+
+```python
+[
+    {
+        "question": "What is the meaning of life?",
+        "reply": "42"
+    },
+    {
+        "question": "What is Hamilton?",
+        "reply": "..."
+    },
+    ...
+]
+```
+
+You would then pass in as _inputs_ to execution `"data_path"=PATH_TO_THIS_FILE` as well as `"input_text_key"="question"` and `"output_text_key"="reply"`.
+- Instantiate the driver. Use `{"start": "base"}` as configuration to run with to use a raw base LLM to finetune.
+- Pick your LLM. `model_id_tokenizer="google/mt5-small"` is the default, but you can change it to any of the models
+that the transformers library supports for `AutoModelForSeq2SeqLM` models.
+- Run the code.
+
+Because there's no configuration that changes the shape of the DAG, you can run the code like this:
+
+```python
+# instantiate the driver with this module however you want
+result = dr.execute(
+    [ # some suggested outputs
+        "save_best_models",
+        "inference_set_predictions",
+        "training_and_validation_set_metrics",
+        "finetuned_model_on_validation_set",
+    ],
+    inputs={
+        "data_path": "example-support-dataset.json", # the path to your dataset
+        "input_text_key": "question",  # the key in the json object that has the input text
+        "output_text_key": "gpt4_replies_target", # the key in the json object that has the target output text
+    },
+)
+```
+
+### Running the code in a docker container
+The easiest way to run this code is to use a docker container, unless you have experience with
+GPUs. After writing a module that uses the code here, e.g. `YOUR_RUN_FILE.py`, you can create a dockerfile
+that looks like this to then execute your fine-tuning code. Note, replace `example-support-dataset.json` with
+your dataset that you want to fine-tune on.
+
+```docker
+FROM python:3.10-slim-bullseye
+
+WORKDIR /app
+
+# install graphviz backend
+RUN apt-get update \
+    && apt-get autoremove -yqq --purge \
+    && apt-get clean \
+    && rm -rf /var/lib/apt/lists/*
+
+COPY requirements.txt .
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY example-support-dataset.json .
+
+COPY . .
+
+EXPOSE 8080:8080
+
+# run the code that you wrote that invokes this module
+CMD python YOUR_RUN_FILE.py
+```
+Then to run this it's just:
+```bash
+docker build -t YOUR_IMAGE_NAME .
+docker run YOUR_IMAGE_NAME
+```
+
+# Configuration Options
+ - `{"start": "base"}` Suggested configuration to run with to use a raw base LLM to finetune.
+ - `{"start":  "presaved"}` Use this if you want to load an already fine-tuned model and then just eval it.
+
+# Limitations
```

Comment:
Scaling up -- we want to point them in the right direction in case they want to use a bigger model/whatnot. Also limitation is that we have some pieces (E.G. modal support) in the works.

Reply:
that's not a limitation of this code.